### PR TITLE
Support `builtin(primitive_index)`

### PR DESCRIPTION
**Pull request by initial-algebra**
Quick and dirty validation and code generation for the SPIR-V builtin `PrimitiveID`.  Currently only attempts to support the SPIR-V backend.

### CODE REVIEW

File: src/back/hlsl/writer.rs [hunk 1]
```diff
@@ -1388,6 +1388,7 @@ fn builtin_str(built_in: crate::BuiltIn) -> &'static str {
         // fragment
         Bi::FragDepth => "SV_Depth",
         Bi::FrontFacing => "SV_IsFrontFace",
+        // Bi::PrimitiveId => "SV_PrimitiveID",
```
Comment (kvark) [inline src/back/hlsl/writer.rs:1391]: why is this commented out?

File: src/lib.rs [hunk 2]
```diff
@@ -269,6 +269,7 @@ pub enum BuiltIn {
     // fragment
     FragDepth,
     FrontFacing,
+    PrimitiveId,
```
Comment (kvark) [inline src/lib.rs:272]: the current naming convention for builtins is such that "XxxID" are vectors, and all the scalars are "XxxIndex".
So let's call this `PrimitiveIndex`.

File: src/valid/mod.rs [hunk 3]
```diff
@@ -58,6 +58,7 @@ bitflags::bitflags! {
         const PUSH_CONSTANT = 0x1;
         /// Float values with width = 8.
         const FLOAT64 = 0x2;
+        const PRIMITIVE_ID = 0x4;
```
Comment (kvark) [inline src/valid/mod.rs:61]: similarly, let's call it `PRIMITIVE_INDEX`

**Review by kvark — COMMENTED** [review]
Great start!

File: src/back/hlsl/writer.rs (hunk 1 above)
Comment (initial-algebra) [reply src/back/hlsl/writer.rs:1391]: I could certainly uncomment it, but I haven't tested the HLSL path at all.  I will do that.  Speaking of backends, I don't have an Apple device or any knowledge of Metal, so I won't be able to work on that platform.  I can certainly throw together a Linux VM to work on the OpenGL backend at some point.

File: src/valid/mod.rs (hunk 3 above)
Comment (initial-algebra) [reply src/valid/mod.rs:61]: Should we maybe call it `GEOMETRY` to reflect the generic shader capability?

Comment (kvark) [reply src/valid/mod.rs:61]: No, we aren't going to support geometry shaders any time soon. From the point of view of Naga, this should be a capability unrelated to geometry shaders. It's just the Naga client's responsibility to check on Vulkan that geometry is allowed.

File: src/back/hlsl/writer.rs (hunk 1 above)
Comment (kvark) [reply src/back/hlsl/writer.rs:1391]: HLSL is emerging now, and we have all the things checked on CI which we care not to break. So don't afraid to be bold on HLSL and MSL. If CI is happy, we are likely happy.
Speaking of which, we'll need to add the use of primitive ID to one of the test. I think "extra.wgsl" should be this, given that it's extra functionality.

**Comment by initial-algebra** [comment]
Very untested, since I'm currently performing SPIR-V passthrough after updating wgpu to `master`.

**Review by kvark — APPROVED** [review]
Shipit!

**Comment by kvark** [comment]
Could you try using it in `extra.wgsl` test and seeing what comes up?

**Comment by initial-algebra** [comment]
Seems to work!

**Comment by kvark** [comment]
@initial-algebra the tests are failing because you didn't add some of the snapshot outputs, it seems

**Review by kvark — APPROVED** [review]

**Comment by initial-algebra** [comment]
I have no idea where to begin with validating the MSL version.

**Comment by kvark** [comment]
Interesting approach! I'll follow up with a way to communicate the language version to the validator, directly.

**Comment by initial-algebra** [comment]
Thank you @kvark for being so patient!

**Pull request merged**